rclcpp/include/rclcpp/client.hpp

```diff
@@ -150,6 +151,69 @@ class ClientBase
   bool
   exchange_in_use_by_wait_set_state(bool in_use_state);
 
+  /// Set a callback to be called when each new response is received.
+  /**
+   * The callback receives a size_t which is the number of responses received
+   * since the last time this callback was called.
+   * Normally this is 1, but can be > 1 if responses were received before any
+   * callback was set.
+   *
+   * Since this callback is called from the middleware, you should aim to make
+   * it fast and not blocking.
+   * If you need to do a lot of work or wait for some other event, you should
+   * spin it off to another thread, otherwise you risk blocking the middleware.
+   *
+   * Calling it again will clear any previously set callback.
+   *
+   * This function is thread-safe.
```

Comment:
@wjwwood I was implementing a way to clear the previously set callbacks and I noticed that the `set_*` functions are not thread safe as mentioned here.

If two threads call `set_on_new_response_callback` concurrently, it may happen that one thread modifies the permanent storage `on_new_response_callback_` while this is being used by the RMW layer.
For example it could happen with the following execution order:
 - thread 1 `set_on_new_response_callback(trampoline, new_callback)`
 - thread 2  `set_on_new_response_callback(trampoline, new_callback)`
 - thread 1 `on_new_response_callback_ = new_callback;`
 - thread 1 `set_on_new_response_callback(trampoline, on_new_response_callback_)` and return 
  - thread 2 `on_new_response_callback_ = new_callback;` ---> this is where problems may arise

Reply:
You're right. We'll have to add a mutex to protect the `on_new_response_callback_` sequence. I wrote the documentation before I changed that code in order to resolve the issue that changing `on_new_response_callback_` would make the old value invalid while it is still set as the callback, allowing for a small window (before ` set_on_new_response_callback` is called) where rcl/rmw could call an invalid callback.

Reply:
Ok, I;m going to add some commits, I will take care of that